MEMORY-MANAGEMENT.md

```diff
@@ -0,0 +1,105 @@
+
+For many of these async operations (such as `setTimeout` and `.then`), a
+callback is run once or multiple times in a task or microtask. In those cases,
+the operation can be seen as keeping a strong reference to the callback, and it
+will also keep a strong reference to the context that was current at the time
+that the API was called to start that operation. When the operation is finished,
+that reference will be removed.
+
+For events, we do not store the context in which `addEventListener` is
+called (though see the next paragraph on the fallback context). Instead, the
+context is propagated from the web API that caused it, if any. For APIs that
+cause events to fire asynchronously (e.g. XHR), this would involve storing a
+reference to the context when the API is called (e.g. `xhr.send()`), and keeping
+it alive until no events can be fired anymore from that asynchronous operation
+(e.g. until the XHR request finishes, errors out or is aborted).
+
+`addEventListener`, however, would need changes with the fallback context
+proposal[^1]. The initial version of this proposal would allow
+`addEventListener` to associate a context with the listener (only if the current
+context contains a key that is set in a call to the fallback context API).
+However, rather than storing a whole context, a newer iteration of this proposal
+(described in https://github.com/tc39/proposal-async-context/issues/107#issuecomment-2659298381)
+would let the API only store the values of one or more `AsyncContext.Variable`s
+that would need to be passed to the API. This means that in practice, this
+context which would be associated to `addEventListener` would only reference one
+or a few objects.
+
+[^1]: This proposal isn't described in any depth in the main web integration
+document because the details are still being worked out. See
+<https://github.com/tc39/proposal-async-context/issues/107>.
+
+The web integration document says that observers (such as MutationObserver,
+IntersectionObserver...) would use the registration context for their callbacks;
+which means when the observer is constructed, it would store a reference to the
+current context, which would never be released while the observer is alive.
+However, it seems like it might be possible to change this behavior so the
+context is not stored at all for observers; instead, the callbacks would be
+called with the empty context.
+
+Although this document and the web integration one describe the context
+propagations that must happen due to the browser and JS engine's involvement,
+it is also important to have in mind how authors might propagate contexts
+implicitly. For example, from the browser's perspective, `requestAnimationFrame`
+only keeps the context referenced until the rAF callback is called. However, the
+context is active when the callback is called, and the callback is likely to
+call `requestAnimationFrame` again, continuing to propagate the same context.
```

Comment:
> For example, from the browser's perspective, `requestAnimationFrame`
only keeps the context referenced until the rAF callback is called.

I didn't get the point after "However, ...". Would you mind elaborating on it?

Reply:
The point was that, although from the perspective of the browser internals, the context from the rAF call only gets propagated to that callback and not to any follow-up rAF callbacks, the author code in the callback itself almost certainly will call rAF without changing the context, which might keep that context alive indefinitely. So that should also be taken into account when considering memory leaks caused by AsyncContext.

Reply:
Reworded to make it clearer.